samza-azure/src/main/java/org/apache/samza/system/eventhub/producer/EventHubSystemProducer.java

```diff
@@ -0,0 +1,337 @@
+    messageId = 0;
+    this.config = config;
+    this.registry = registry;
+    partitioningMethod = this.config.getPartitioningMethod();
+    this.eventHubClientWrapperFactory = eventHubClientWrapperFactory;
+    this.serdes = serdes;
+  }
+
+  @Override
+  public synchronized void start() {
+    LOG.info("Starting system producer.");
+    for (String eventHub : eventHubClients.keySet()) {
+      eventWriteRate.put(eventHub, registry.newCounter(eventHub, EVENT_WRITE_RATE));
+      eventByteWriteRate.put(eventHub, registry.newCounter(eventHub, EVENT_BYTE_WRITE_RATE));
+      sendLatency.put(eventHub, new SamzaHistogram(registry, eventHub, SEND_LATENCY));
+      sendCallbackLatency.put(eventHub, new SamzaHistogram(registry, eventHub, SEND_CALLBACK_LATENCY));
+      sendErrors.put(eventHub, registry.newCounter(eventHub, SEND_ERRORS));
+    }
+
+    // Locking to ensure that these aggregated metrics will be created only once across multiple system consumers.
+    synchronized (AGGREGATE) {
+      if (aggEventWriteRate == null) {
+        aggEventWriteRate = registry.newCounter(AGGREGATE, EVENT_WRITE_RATE);
+        aggEventByteWriteRate = registry.newCounter(AGGREGATE, EVENT_BYTE_WRITE_RATE);
+        aggSendLatency = new SamzaHistogram(registry, AGGREGATE, SEND_LATENCY);
+        aggSendCallbackLatency = new SamzaHistogram(registry, AGGREGATE, SEND_CALLBACK_LATENCY);
+        aggSendErrors = registry.newCounter(AGGREGATE, SEND_ERRORS);
+      }
+    }
+
+    isStarted = true;
+  }
+
+  @Override
+  public synchronized void stop() {
+    LOG.info("Stopping event hub system producer...");
+    streamPartitionSenders.values().forEach((streamPartitionSender) -> {
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+        streamPartitionSender.forEach((key, value) -> futures.add(value.close()));
+        CompletableFuture<Void> future = CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+        try {
+          future.get(config.getShutdownWaitTimeMS(), TimeUnit.MILLISECONDS);
+        } catch (ExecutionException | InterruptedException e) {
+          LOG.warn("Closing the partition sender failed ", e);
+        } catch (TimeoutException e) {
+          LOG.warn("Closing the partition sender timed out ", e);
+        }
+      });
+    eventHubClients.values().forEach(ehClient -> ehClient.close(config.getShutdownWaitTimeMS()));
+    eventHubClients.clear();
+  }
+
+  @Override
+  public synchronized void register(String streamName) {
+    LOG.info("Trying to register {}.", streamName);
+    if (isStarted) {
+      String msg = "Cannot register once the producer is started.";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    EventHubClientWrapper ehClient = eventHubClientWrapperFactory
+            .getEventHubClientWrapper(config.getStreamNamespace(streamName), config.getStreamEntityPath(streamName),
```

Comment:
Extract variables.

Formatting: There are several instances where we're inlining verbose method calls and line wrapping half-way through the statement. IMHO it's much more readable to extract variables instead.